### PR TITLE
[SPARK-12010][SQL] Spark JDBC requires support for column-name-free INSERT syntax

### PR DESCRIPTION
**Pull request by CK50**
This PR moves the generation of the SQL Insert Statement into the JdbcDialect and allows dialects to add column names based on the RDD column names. The default behaviour stays unchanged. Only for the new dialect ProgressCassandraDialect column names will be inserted into the SQL Insert Statement.

I was wondering whether the logic for inserting the column names should be placed in some kind of generic Dialect instead of into the very specific ProgressCassandraDialect. Any suggestions welcome.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/ProgressCassandraDialect.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package org.apache.spark.sql.jdbc
+
+import java.sql.Types
+
+import org.apache.spark.sql.types._
+
+
+private case object ProgressCassandraDialect extends JdbcDialect {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/ProgressCassandraDialect.scala]: what does progress mean here?

**Comment by rxin** [comment]
@CK50 this needs some style cleanup - you can run dev/lint-scala to check styles locally.

Can you please also attach the generated query before / after this change? IIUC, maybe it's ok to have this for all dialects.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/ProgressCassandraDialect.scala (hunk 1 above)
Comment (CK50) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/ProgressCassandraDialect.scala]: Hi Reynold,
Progress is a software vendor. They are/will be offering a JDBC driver for Cassandra.
Thanks,
Christian

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/ProgressCassandraDialect.scala]: Are there other Cassandra jdbc drivers and is this the primary one people use? I'm thinking we should just drop "Progress" here.

Comment (CK50) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/ProgressCassandraDialect.scala]: Yes, there is a bunch of other Cassandra JDBC drivers out there. Progress one of them. But seen that the Cassandra SQL syntax mandates the list of column names in the INSERT statement, it is likely that other drivers will require the same syntax. So I will remove "Progress" from the class name.

**Comment by srowen** [comment]
This is an orthogonal issue, but it occurs to me that we have a bunch of dialects, and a bunch of behavior variations, but typically a variation isn't specific to a dialect, necessarily. How do we let a new variation apply to N dialects at once? this may become a problem soon and need some refactoring.

**Comment by CK50** [comment]
@rxin
I wish I could run dev/lint-scala, but even after hours of struggling I 
cannot get build/sbt running. I have downloaded sbt-launcher.jar 
manually, but now I am stuck with sbt trying to download extra files 
from a variety of invalid urls such as
https://jcenter.bintray.com/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.pom
https://jcenter.bintray.com/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.jar
Any ideas appreciated. The Spark 1.6 build page did not help. - Any 
other page I should follow?

The old generated statement is

INSERT INTO mytable VALUES (?, ?, ..., ?)

whereas the new statement is

INSERT INTO mytable (col1, col2,  ..., colN) VALUES (?, ?, ..., ?)

So the old syntax relies on the column positions, whereas the new one 
relies on the column names.

Column names are taken from the DataFrame. So DataFrame column names 
must match target table column names.

At least for Oracle adding the column names is fine. The benefit of this 
change is that you can write out DataFrames with having less columns 
than the target table, which I think is not possible today. The downside 
is that DataFrame column names *must* match. For best backwards 
compatibility I only wanted to provide column names when it is really 
needed, like in case of Cassandra.

WDYT?

**Comment by srowen** [comment]
I'm not sure what the issue is with SBT (do you get download errors or something?) but you can also see the results of the style checker in the build output: https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2119/consoleFull
Search for "[error]"

I tend to think it's better practice to specify column names explicitly. Is there a downside? I don't recall a database that _wouldn't_ let you specify them, right?

**Comment by CK50** [comment]
Do I take this as "SBT works fine for me on master"?

Yes, build/sbt fails for me when doing downloads:

Getting org.scala-sbt sbt 0.13.7 ...

:: problems summary ::
:::: WARNINGS
        module not found: org.scala-sbt#sbt;0.13.7
   ==== local: tried
     /home/ckurz/.ivy2/local/org.scala-sbt/sbt/0.13.7/ivys/ivy.xml
     -- artifact org.scala-sbt#sbt;0.13.7!sbt.jar:
     /home/ckurz/.ivy2/local/org.scala-sbt/sbt/0.13.7/jars/sbt.jar
   ==== jcenter: tried
https://jcenter.bintray.com/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.pom
     -- artifact org.scala-sbt#sbt;0.13.7!sbt.jar:
https://jcenter.bintray.com/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.jar
   ==== typesafe-ivy-releases: tried
https://repo.typesafe.com/typesafe/ivy-releases/org.scala-sbt/sbt/0.13.7/ivys/ivy.xml
   ==== Maven Central: tried
https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.pom
     -- artifact org.scala-sbt#sbt;0.13.7!sbt.jar:
https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.jar
        ::::::::::::::::::::::::::::::::::::::::::::::
        ::          UNRESOLVED DEPENDENCIES         ::
        ::::::::::::::::::::::::::::::::::::::::::::::
        :: org.scala-sbt#sbt;0.13.7: not found
        ::::::::::::::::::::::::::::::::::::::::::::::

:::: ERRORS
   Server access Error: Connection timed out 
url=https://jcenter.bintray.com/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.pom
   Server access Error: Connection timed out 
url=https://jcenter.bintray.com/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.jar
   Server access Error: Connection timed out 
url=https://repo.typesafe.com/typesafe/ivy-releases/org.scala-sbt/sbt/0.13.7/ivys/ivy.xml
   Server access Error: Connection timed out 
url=https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.pom
   Server access Error: Connection timed out 
url=https://repo1.maven.org/maven2/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.jar


I am behind a proxy and have tried configuring linux env vars and 
JAVA_OPTS/SBT_OPTS, but this did not help.

The thing, which strikes me is that when I try accessing these urls: 
https://jcenter.bintray.com/org/scala-sbt/sbt/0.13.7/sbt-0.13.7.pom
from within a browser, I also get 404. When looking into 
https://jcenter.bintray.com/org/scala-sbt there is no sub-folder called 
sbt. - No idea what is going wrong here :-(


----------------------
Back to the PR:
I will upload the suggested style changes.

In general I agree that column names should be given by default. I just 
have two concerns re making this the new default behaviour:

1) does this work on all supported DBs? - Does Spark have automated 
tests with appropriate envs for this?
2) is it okay to assume that DataFrame column names always match target 
col names? (Spark currently does not rely on this. Instead it relies on 
column positions, which is usually not good practice.)
3) is such a change okay with Spark's rules for backwards compatibility?

Just some concerns, I am fine with making column names the new default 
behaviour. But as I said, I do not have the background to judge this.

Please let me know how to proceed.

**Comment by srowen** [comment]
It worked fine for me locally since I have `sbt` available locally. I tried it on another machine so that it has to download sbt. Although it seemed to take some time, it succeeded. It looks maybe like a transient problem accessing the repos? I didn't seem to connect to bintray for some reason, but indeed those URLs look like they don't work. But they're not in our build. Hm. Try again?

(This will need a rebase/force push BTW)

**Comment by CK50** [comment]
I have tried to rebase this PR to the lastest of apache/spark. - Does this look like expected?

**Comment by srowen** [comment]
No, this branch has a lot of unrelated commits now. Generally you keep your fork's `master` up to date with the main one, and then brings yours up to date, then switch to your branch and `git rebase master`.

You can hard-reset back over the bad commits or just start over.

**Comment by srowen** [comment]
@CK50 do you mind closing this PR? maybe easier than rebasing and untangling it.

**Comment by CK50** [comment]
Need to create a new pull request for this fix.

**Pull request closed without merging**